module/x/peggy/keeper/keeper.go

```diff
@@ -136,6 +155,57 @@ func (k Keeper) GetValsets(ctx sdk.Context) (out []*types.Valset) {
 	return
 }
 
+// GetLatestValset returns the latest validator set in state
+func (k Keeper) GetLatestValset(ctx sdk.Context) (out *types.Valset) {
+	latestValsetNonce := k.GetLatestValsetNonce(ctx)
+	out = k.GetValset(ctx, latestValsetNonce)
+	return
+}
+
+// setLastSlashedValsetNonce sets the latest slashed valset nonce
+func (k Keeper) SetLastSlashedValsetNonce(ctx sdk.Context, nonce uint64) {
+	store := ctx.KVStore(k.storeKey)
+	store.Set(types.LastSlashedValsetNonce, types.UInt64Bytes(nonce))
+}
+
+// GetLastSlashedValsetNonce returns the latest slashed valset nonce
+func (k Keeper) GetLastSlashedValsetNonce(ctx sdk.Context) uint64 {
+	store := ctx.KVStore(k.storeKey)
+	bytes := store.Get(types.LastSlashedValsetNonce)
+
+	if len(bytes) == 0 {
+		return 0
+	}
+	return types.UInt64FromBytes(bytes)
+}
+
+// GetUnSlashedValsets returns all the unslashed validator sets in state
+func (k Keeper) GetUnSlashedValsets(ctx sdk.Context) (out []*types.Valset) {
+	lastSlashedValsetNonce := k.GetLastSlashedValsetNonce(ctx)
+	latestValsetNonce := k.GetLatestValsetNonce(ctx)
+	k.IterateValsetBySlashedValsetNonce(ctx, lastSlashedValsetNonce, latestValsetNonce, func(_ []byte, valset *types.Valset) bool {
```

Comment:
We should instead just iterate up to the `uint64(ctx.BlockHeight())-params.SignedValsetsWindow` block height,